src/dmd/dsymbolsem.d

```diff
@@ -569,6 +569,29 @@ const(char)* getMessage(DeprecatedDeclaration dd)
     return dd.msgstr;
 }
 
+// Returns true if a contact can appear with a function body.
+private bool allowsContractWithoutBody(FuncDeclaration funcdecl)
+{
+    if (!funcdecl.fbody)
+    {
+        /* Contracts can only appear without a body when they are virtual
+         * interface functions or abstract.
+         */
+        Dsymbol parent = funcdecl.toParent();
+        InterfaceDeclaration id = parent.isInterfaceDeclaration();
+
+        if (!funcdecl.isAbstract() &&
+            (funcdecl.fensure || funcdecl.frequire) &&
+            !(id && funcdecl.isVirtual()))
+        {
+            auto cd = parent.isClassDeclaration();
+            if (!(cd && cd.isAbstract()))
+                return false;
+        }
```

Comment:
+1
This is a lot better in terms of readability!

Reply:
Actually, I think the readability is impaired by the extensive use of `!`. Not a big deal, but I'd De Morgan the sense around so it matches the comment.